Typescript

glenngijsberts opened the pull request:
# Description

This PR will introduce Typescript!

## How to contribute

* Checkout PR
* Make sure you are on the latest commit
* Do not force push so we can all safely use `git pull`

**Component**

* Change component to `.tsx`
* Make sure to type the file. Use `enum` types as much as possible
* Export Typescript related stuff before declaration like: `export interface AlertProps`
* Export components at the bottom of the file like: `export { Accordion, AccordionPanel ... }`

**Storybook**

* Change storybook file to `.tsx` and fix potential typings such as applying enums
* Export stories at the bottom as much as possible

**Tests** 
* Change test file to `.tsx` if it contains jsx code, otherwise change it to `.ts`
* Fix the tests if needed, such as applying enums

For inspiration and/or examples you can take a look at components that are already done such as `Alert` & `Accordion`. The goal is refactoring the code to TypeScript, without making changes to the functionality. However, if you may need to remove some code or do need to change functionality you should mark this as a breaking change in your commit, using `yarn commit`.

## Changes

- [x] Enable typescript
- [x] Generate declarations
- [x] Enable Typescript for Jest tests 
- [x] Hooks - Typescript ready
- [x] Utils - Typescript ready
- [x] Use TSDX

- [x] Fix strict warning and errors


- [x] Components:
- - [x] Alert
- - [x] Avatar
- - [x] Accordion
- - [x] Banner
- - [x] Text
- - [x] BaseStyles
- - [x] StarRating
- - [x] Button
- - [x] Card
- - [x] ContentDialog
- - [x] Panel
- - [x] Pill
- - [x] BaseButton
- - [x] Collapsible
- - [x] Toggle
- - [x] Checkbox
- - [x] Select
- - [x] DateInput
- - [x] MoneyInput
- - [x] PhoneInput
- - [x] TimeInput
- - [x] Combobox
- - [x] ContentTransition
- - [x] Skeleton
- - [x] Toast
- - [x] Cover
- - [x] CustomDropdown
- - [x] Dialog
- - [x] OnboardingDialog
- - [x] Flag
- - [x] global-styles
- - [x] Heading
- - [x] HorizontalScroll
- - [x] Image
- - [x] Input
- - [x] Logo
- - [x] MenuButton
- - [x] Portal
- - [x] SpeechBubble
- - [x] Spinner
- - [x] Tabs 
- - [x] VisuallyHidden
- - [x] Tooltip

## Breaking changes

* `easingFunctions` are removed from `theme.js`
* `debounce` is removed
* The `handler` parameter of `useOnClickOutside` is no longer exposing the `event` after firing

glenngijsberts committed:
refactor: Use typescript for components

glenngijsberts committed:
build: Use TSDX for bundle setup

glenngijsberts committed:
perf(easingfunctions): easingFunctions are no longer used and therefore being removed

BREAKING CHANGE: the easingFunctions are removed from the theme file.

glenngijsberts committed:
build: add Typescript

glenngijsberts committed:
perf(debounce): remove debounce util

The debounce util is no longer used and therefore will be removed from the package

BREAKING CHANGE: Debounce util is no longer used. Please use another alternative such as debounce
from lodash

glenngijsberts committed:
build(jest): make the tests work with Typescript

glenngijsberts committed:
perf(useonclickoutside): use Typescript and removed unused event argument for handler parameter

RobVermeer commented on src/components/Pill/index.tsx:
```diff
@@ -46,29 +46,18 @@ const LeftAdornment = styled.span`
-    </Container>
-  )
+interface PillProps {
+  variant: String
```
I think you can do the following, so it's an enum list of variants.

```typescript
enum Variant {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  danger = 'danger',
  warning = 'warning',
  disabled = 'disabled',
}

interface PillProps {
  variant: Variant,
  leftAdornment: Element,
}
```

Commit:
build: Add eslint typescript support

glenngijsberts committed:
build(jest): lower the treshhold for functions

Currently tests for our components are not up to date. We lower the treshold for now, but wewill
look into updating our tests in the future so we can increase the treshold even higher in the future

glenngijsberts committed:
refactor(skeleton): use Typescript; removed theme prop

The theme prop is removed; you can change the liniear-gradient using styled components. Also changed
the exstension to typescript.

BREAKING CHANGE: Theme prop is removed; use styled components overwriting instead

rppld commented:
Check out [TSDX](https://github.com/formium/tsdx). Might simplify the bundling setup quite a bit.

rppld commented on src/components/Alert/index.tsx:
```diff
@@ -19,23 +18,41 @@ import {
+  onClick: (event: React.MouseEvent<HTMLElement>) => void
+}
+
+export enum AlertVariant {
+  info = 'info',
+  success = 'success',
+  error = 'error',
+  warning = 'warning',
+}
```
There’s a [case against](https://blog.logrocket.com/why-typescript-enums-suck/) using enums in TypeScript. Might be clearer and less error-prone to use string union types.
```suggestion
export type AlertVariant = 'info' | 'succes' | 'error' | 'warning'
```

glenngijsberts replied:
Thanks Philipp! Let's discuss tomorrow @RobVermeer @Roohn ;)

glenngijsberts committed:
build: change prepare script name to prepublishOnly

RobVermeer commented:
:tada: This PR is included in version 14.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@ticketswap/solar/v/14.0.0)
- [GitHub release](https://github.com/TicketSwap/solar/releases/tag/v14.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
